[Messenger] Added documentation about DispatchAfterCurrentBusMiddleware

Nyholm opened the pull request:
Documentation to PR: https://github.com/symfony/symfony/pull/27844













![Screenshot 2019-03-17 at 11 43 23](https://user-images.githubusercontent.com/1275206/54489171-edee4880-48a9-11e9-8028-ac501e62dfcb.png)
![Screenshot 2019-03-17 at 11 43 34](https://user-images.githubusercontent.com/1275206/54489173-edee4880-48a9-11e9-877f-f749f242b502.png)

![Screenshot 2019-03-17 at 11 43 39](https://user-images.githubusercontent.com/1275206/54489172-edee4880-48a9-11e9-84f1-d78e68474c7a.png)

Nyholm committed:
Added documentation about message recorder

javiereguiluz commented:
@Nyholm thanks for this contribution! Lately in the Symfony Docs we're trying to not add new articles when they are too short. So, do you think we could move this article to a section inside an existing article? If there is no article where we can put this ... or if such article is already super long, we could consider creating this new article. Thanks!

Nyholm commented:
Thank you. I agree.

I did not want my PR's to build on each other. At the same time, the message recorder has not been merged yet (Im not sure that it will). 

#10013 is blocking. When #10013 is merged I will put this text into that article.

HeahDude commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,69 @@
+
+    .. code-block:: yaml
+
+        # config/packages/workflow.yaml
```
```diff
- workflow.yaml
+ messenger.yaml
```

HeahDude commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,69 @@
+Record Events Produced by a Handler
+===================================
+
+In a example application there is a command (a CQRS message) named ``CreateUser``.
```
"In an example"

HeahDude commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,69 @@
+===================================
+
+In a example application there is a command (a CQRS message) named ``CreateUser``.
+That command is handled by the ``CreateUserHandler``. The command handler creates
```
```diff
- by the ``CreateUserHandler``. The command handler creates
+ by the ``CreateUserHandler`` which creates
```
?

HeahDude commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,69 @@
+
+In a example application there is a command (a CQRS message) named ``CreateUser``.
+That command is handled by the ``CreateUserHandler``. The command handler creates
+a ``User`` object, stores that object to a database and dispatches an ``UserCreatedEvent``.
```
"a UserCreatedEvent"

Nyholm replied:
Are you sure?

HeahDude replied:
Almost :) I'm not native. But you wrote "a User" just before. See https://english.stackexchange.com/questions/105116/is-it-a-user-or-an-user

Nyholm replied:
I would say that "a User" was a typo. 

I read your source. I did not know that. Thank you. I've updated the PR.

HeahDude commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,69 @@
+There are many subscribers to the ``UserCreatedEvent``, one subscriber may send
+a welcome email to the new user. Since we are using the ``DoctrineTransactionMiddleware``
+we wrap all database queries in one database transaction and rollback that transaction
+if an exception is thrown. That would mean that if an exception is thrown when sending
```
```diff
That would mean
That means
```

HeahDude commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,69 @@
+                            - messenger.middleware.validation
+                            - messenger.middleware.handles_recorded_messages: ['@messenger.bus.event']
+                              # Doctrine transaction must be after handles_recorded_messages middleware
+                            - app.doctrine_transaction_middleware: ['default']
```
What about using an FQCN for `app.doctrine_transaction_middleware`? or may we explicit the definition somewhere. If possible I'd rather use the class name.

Nyholm replied:
I have an idea that this will be on the same page as #10013 where we use `app.doctrine_transaction_middleware`.

HeahDude commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,69 @@
+
+.. code-block:: php
+
+    use Doctrine\ORM\EntityManagerInterface;
```
We're missing a namespace here. `App\Messenger\Handler`?

Nyholm replied:
Thank you. Since the example has multiple buses, I prefer `App\Messenger\CommandHandler`.

HeahDude commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,69 @@
+            $this->em = $em;
+        }
+
+        public function __invoke(CreateUser $command)
```
We're missing use statements for domain classes, what about:
```php
use App\Entity\User;
use App\Messenger\Command\CreateUser;
use App\Messenger\Event\UserCreatedEvent;
```
?

Nyholm replied:
Thank you

Nyholm committed:
According to feedback

Nyholm committed:
a user

sroze commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,74 @@
+Record Events Produced by a Handler
+===================================
+
+In an example application there is a command (a CQRS message) named ``CreateUser``.
```
```patch
- In an example application there is a command (a CQRS message) named ``CreateUser``.
+ Let's take the example of an application that has a command (a CQRS message) named ``CreateUser``.
```

?

sroze commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,74 @@
+
+In an example application there is a command (a CQRS message) named ``CreateUser``.
+That command is handled by the ``CreateUserHandler`` which creates
+a ``User`` object, stores that object to a database and dispatches a ``UserCreatedEvent``.
```
``UserCreated`` event?

sroze commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,74 @@
+That event is also a normal message but is handled by an *event* bus.
+
+There are many subscribers to the ``UserCreatedEvent``, one subscriber may send
+a welcome email to the new user. Since we are using the ``DoctrineTransactionMiddleware``
```
I'd suggest to split the two and explicit the "problem":

```patch
- Since we are using the ``DoctrineTransactionMiddleware`` [...]
+ We are using the ``DoctrineTransactionMiddleware`` to wrap all database queries in one database transaction.
+
+ **Problem:** if an exception is thrown when sending the welcome email, then the user will not be created because the ``DoctrineTransactionMiddleware`` will rollback the Doctrine transaction, in which the user has been created.
```

Nyholm replied:
Nice. Thank you

sroze commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,74 @@
+if an exception is thrown. That means that if an exception is thrown when sending
+the welcome email, then the user will not be created.
+
+The solution to this issue is to not dispatch the ``UserCreatedEvent`` in the
```
Same, I'd highlight the solution:
```
**Solution:**
```

sroze commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,74 @@
+.. index::
+    single: Messenger; Record messages
+
+Record Events Produced by a Handler
```
Might be better to find a name that describes the problem than the technical solution. Something like "Event recorder: Tolerating failures while in Doctrine transactions"?

sroze commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,74 @@
+            $user = new User($command->getUuid(), $command->getName(), $command->getEmail());
+            $this->em->persist($user);
+
+            $this->eventRecorder->record(new UserCreatedEvent($command->getUuid());
```
You need to explain what this does, right? Because it does more than "recording" the event. Actually, this call does _postpone the event dispatching_ isn't it?

Nyholm replied:
The EventRecorder is just a storage. There is a middleware that reads from that store later. But yes, I'll write something

Nyholm committed:
Added fixes

Nyholm commented:
Thank you for the review. I've updated the PR accordingly

arnolanglade commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,77 @@
+Events Recorder: Handle Events After CommandHandler Is Done
+===========================================================
+
+Let's take the example of an application that has a command (a CQRS message) named
```
This is not `a CQRS message` but message. You suer messages too when you application use the CQS pattern for example.

magnusnordlander replied:
Indeed, CQRS doesn't necessitate any form of messages. Could "an imperative domain message" be an option? Or maybe a more thorough definition could be useful.

arnolanglade commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,77 @@
+===========================================================
+
+Let's take the example of an application that has a command (a CQRS message) named
+``CreateUser``. That command is handled by the ``CreateUserHandler`` which creates
```
It could be better to use `SignUpUser` as command name, it is less focus on CRUD action.

arnolanglade commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,77 @@
+        public function __invoke(CreateUser $command)
+        {
+            $user = new User($command->getUuid(), $command->getName(), $command->getEmail());
+            $this->em->persist($user);
```
Recording/persisting should be done by the user repository.

arnolanglade replied:
Or a middleware :)

sroze replied:
_can_ be done by either. It works this way as well :)

arnolanglade commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,77 @@
+            $this->em->persist($user);
+
+            // "Record" this event to be processed later by "handles_recorded_messages".
+            $this->eventRecorder->record(new UserCreatedEvent($command->getUuid());
```
This event should be produced by the user model.

arnolanglade reviewed (commented):
It would be good to add a rich model `User` with a named constructor `signUp` that will recored a `SignedUpUser` event.

magnusnordlander commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,77 @@
+``CreateUserHandler`` but to just "record" the events. The recorded events will
+be dispatched after ``DoctrineTransactionMiddleware`` has committed the transaction.
+
+To enable this, you simply just add the ``messenger.middleware.handles_recorded_messages``
```
We should remove "simply just", because it's just filler, and not everyone would consider this simple.

gubler commented:
The current PR symfony/symfony#28849 replaces the RecordsMessages middleware and recorder with a `HandleMessageInNewTransaction` middleware. I have did a rewrite of the documentation but I'm not sure how to request a merge into an existing merge request.

My rewrite is here: https://github.com/gubler/messenger-transaction-test/blob/master/docs/transactional-messages.rst

I tried to take into account previous comments on this request. Feel free to take whatever is useful from my rewrite.

Nyholm commented:
Thank you @gubler Could you open a PR to my [fork](https://github.com/Nyholm/symfony-docs/tree/message-recorder)?

gubler committed:
Explain HandleMessageInNewTransaction middleware

The current PR is for HandleMessageInNewTransaction middleware instead of the original RecordsMessages middleware. This documentation covers the new middleware.

gubler commented:
Done - though its the first time I've done this, so I apologize if I messed anything up :/

ogizanagi commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,120 @@
+                        middleware:
+                            - validation
+                            - handle_message_in_new_transaction
+                            - doctrine_transaction
```
doctrine transaction middleware in the event bus isn't a necessity. I would not add it here as it's only used in the command bus to demonstrate the events are only dispatched after the doctrine transaction is committed.
The event may not even use doctrine, not have a handler, or could just be a message sent asynchronously to a queue. So what produces the event here is another matter and the `doctrine_transaction` here might only bring confusion to the reader.

I'm not sure either about validation, as you're the one creating the event, I guess this would be a logic error from the developper if the crafted event is invalid somehow.

gubler replied:
That makes sense - even if they may be useful in some people's applications, there is no reason to have them in example code.

- The middleware order for the command bus shows that the `handle_message_in_new_transaction` must be before `doctine_transaction`
- The `handle_message_in_new_transaction` being declared for both buses shows that it must be declared of all involved buses.

Since both notes/warning are shown in the example, no reason to have extra middlewares.

```suggestion
                        middleware:
                            - handle_message_in_new_transaction
```

should be sufficient for the example

weaverryan replied:
add it commented out with a description above:

```
> # uncomment if you want each message to be wrapped in a transaction
> # when used with handle_message_in_new_transaction, each message will be wrapped
> # in their own, separate transactions
```

sroze replied:
I agree, we could remove it.

weaverryan commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,120 @@
+  and an exception is thrown, then any database transactions in the original handler will
+  be rolled back.
+- If the message is dispatched to a different bus, then the dispatched message can still
+  be handled even if the original handler encounters an exception.
```
"can still be handled"? Are you saying "the dispatched message will still be handled (which is good) even if the original handler has an exception" or "the dispatched message will still be handled (which is bad) even if the original handler has an exception". I can't tell what the current behavior vs desired behavior is.

weaverryan commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,120 @@
+
+**Problem 1:** If an exception is thrown when sending the welcome email, then the user
+will not be created because the ``DoctrineTransactionMiddleware`` will rollback the
+Doctrine transaction, in which the user has been created.
```
I thought (from your description earlier) that because `UserSignedUp` is sent to a different bus, if it throws an exception, it will not affect the transaction wrapping the command bus... and so that transaction (the one that saves the user) will still be committed. I'm sure i'm wrong - but I'm getting a bit lost :)

sroze replied:
Nop, that's the point of this `DispatchAfterCurrentBusStamp` :)

weaverryan commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,120 @@
+            $this->eventBus->dispatch(
+                (new Envelope($event))
+                    ->with(new Transaction())
+            );
```
This is a bit confusing: why/when would this handler throw an exception? This is the last line of this function... so if we get there, how would an exception occur? And where should the flush live?

sroze replied:
Indeed, at this point is should be okay... unless there is a middleware doing so crazy things 😉

Nyholm commented:
We should not forget this: https://github.com/symfony/symfony/pull/28849#discussion_r265639129

Nyholm committed:
Minor updates

Nyholm commented:
I did some updates to the docs to keep up-to-date with the symfony PR. 

I also added some screenshots to this PR description. I hope they will address any questions about the functionality and to help you (everyone) to help me explain this in the docs.

sroze commented:
Thank you @Nyholm!

Nyholm commented:
@sroze could you add your review?

Nyholm committed:
Added PHP and XML config

OskarStark commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,188 @@
+.. index::
+    single: Messenger; Record messages; Transaction messages
+
+Transactional Messages: Handle Events After CommandHandler Is Done
```
```suggestion
Transactional Messages: Handle Events After CommandHandler is Done
```

OskarStark commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,188 @@
+- If the message is dispatched to a different bus, then the dispatched message can still
+  be handled even if the original handler encounters an exception.
+
+An Example ``SignUpUser`` Process
```
SignUpUser looks strange, can we use RegisterUser for this example?

OskarStark commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,188 @@
+
+    .. code-block:: xml
+
+        <!-- config/packages/cache.xml -->
```
```suggestion
        <!-- config/packages/messenger.xml -->
```

Nyholm replied:
Oops :)

OskarStark commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,188 @@
+
+    .. code-block:: php
+
+        // config/packages/cache.php
```
```suggestion
        // config/packages/messenger.php
```

OskarStark commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,188 @@
+
+    class SignUpUserHandler
+    {
+        private $em;
```
the member vars should have the same order, like the constructor args

OskarStark commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,188 @@
+
+    use App\Entity\User;
+    use App\Messenger\Event\UserSignedUp;
+    use Doctrine\ORM\EntityManagerInterface;
```
`MyMailer` use statement missing. Maybe you could use symfony/mailer ?

Nyholm replied:
Thanks!

OskarStark commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,188 @@
+            $this->em = $em;
+        }
+
+        public function __invoke(UserSignedUp $eent)
```
```suggestion
        public function __invoke(UserSignedUp $event)
```

OskarStark commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,188 @@
+        }
+    }
+
+**Note:** If ``WhenUserSignedUpThenSendWelcomeEmail`` throws an exception, that exception
```
please use ´.. note::´ instead

OskarStark committed:
Update messenger/message-recorder.rst

Co-Authored-By: Nyholm <tobias.nyholm@gmail.com>

OskarStark committed:
Update messenger/message-recorder.rst

Co-Authored-By: Nyholm <tobias.nyholm@gmail.com>

OskarStark committed:
Update messenger/message-recorder.rst

Co-Authored-By: Nyholm <tobias.nyholm@gmail.com>

OskarStark committed:
Update messenger/message-recorder.rst

Co-Authored-By: Nyholm <tobias.nyholm@gmail.com>

Nyholm committed:
updates according to feedback

OskarStark commented on messenger/message-recorder.rst:
```diff
@@ -0,0 +1,192 @@
+.. index::
+    single: Messenger; Record messages; Transaction messages
+
+Transactional Messages: Handle Events After CommandHandler is Done
```
@javiereguiluz s/After/after/ ?

sroze reviewed (approved):
Except these few details, it looks good to me 👍

Nyholm committed:
minor updates according to feedback

Nyholm commented:
Thank you for the review. I've updated accordingly

weaverryan commented:
Thank you @Nyholm and @gubler!

Nyholm commented:
Thank you for merging.
